src/zarr/buffer.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import sys
+from typing import TYPE_CHECKING, Any, Callable, Iterable, Literal, Optional, Tuple
+import numpy as np
+
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+    from zarr.codecs.bytes import Endian
+
+
+class NDBuffer:
+    """A n-dimensional memory block
+
+    We use `NDBuffer` throughout Zarr to represent a block of memory.
+    For now, we only support host memory but the plan is to support other types
+    of memory such as CUDA device memory.
+    """
+
+    def __init__(self, array: np.ndarray):
+        assert isinstance(array, np.ndarray)
+        assert array.dtype != object
+        self._data = array
+
+    @classmethod
+    def create_empty(
+        cls, *, shape: Iterable[int], dtype: np.DTypeLike, order: Literal["C", "F"] = "C"
+    ) -> Self:
+        return cls(np.empty(shape=shape, dtype=dtype, order=order))
+
+    @classmethod
+    def create_zeros(
+        cls, *, shape: Iterable[int], dtype: np.DTypeLike, order: Literal["C", "F"] = "C"
+    ) -> Self:
+        return cls(np.zeros(shape=shape, dtype=dtype, order=order))
+
+    def as_numpy_array(self, dtype: Optional[np.DTypeLike] = None) -> np.ndarray:
+        if dtype is None:
+            return self._data
+        return self._data.astype(dtype=dtype, copy=False)
+
+    @property
+    def dtype(self) -> np.dtype[Any]:
+        return self.as_numpy_array().dtype
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        return self.as_numpy_array().shape
+
+    @property
+    def byteorder(self) -> Endian:
+        from zarr.codecs.bytes import Endian
+
+        if self.dtype.byteorder == "<":
+            return Endian.little
+        elif self.dtype.byteorder == ">":
+            return Endian.big
+        else:
+            return Endian(sys.byteorder)
+
+    def reshape(self, newshape: Iterable[int]) -> Self:
+        return self.__class__(self.as_numpy_array().reshape(newshape))
+
+    def astype(self, dtype: np.DTypeLike, order: Literal["K", "A", "C", "F"] = "K") -> Self:
+        return self.__class__(self.as_numpy_array().astype(dtype=dtype, order=order))
+
+    def __getitem__(self, key: Any) -> Self:
+        return self.__class__(np.asanyarray(self.as_numpy_array().__getitem__(key)))
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        if isinstance(value, NDBuffer):
+            value = value.as_numpy_array()
+        self.as_numpy_array().__setitem__(key, value)
+
+    def __len__(self) -> int:
+        return self.as_numpy_array().__len__()
+
+    def fill(self, value: Any) -> None:
+        self.as_numpy_array().fill(value)
+
+    def copy(self) -> Self:
+        return self.__class__(self.as_numpy_array().copy())
+
+    def transpose(self, *axes: np.SupportsIndex) -> Self:
+        return self.__class__(self.as_numpy_array().transpose(*axes))
+
+
+class Buffer(NDBuffer):
```

Comment:
Why is the `Buffer` subclassing `NDBuffer`? I suppose the API of the `Buffer` would look more like `bytes` than `ndarray`.

Reply:
To show that a component that supports `NDBuffer` also supports `Buffer` but not necessarily the other way around. 
A `Buffer` is just a specialized instantiation of a `NDBuffer` with `dim=1, stride=1, dtype="b"`.

Reply:
What components do you have in mind that would support both an `NDBuffer` and a flat `Buffer`?
In zarr-python 3 we are quite deliberate about turning ndarrays into bytes and vice versa. Afaik this only happens in the `ArrayBytesCodec`s. I think that is a big win in terms of clarity. I would prefer to keep the buffer types separate to not make it too easy for components to do implicit conversions on their own.

Reply:
> What components do you have in mind that would support both an `NDBuffer` and a flat `Buffer`?

All components implementing one of:
```
ArrayArrayCodec
ArrayArrayCodec
ArrayBytesCodec
ArrayBytesCodecPartialDecodeMixin
ArrayBytesCodecPartialEncodeMixin
``` 
But I see you point, it is a good idea to separate the classes to avoid implicit conversion. 
However, do we have a good reason to support non-contiguous arrays? If not, I would strongly suggest that we use `NDBuffer` exclusively; components that doesn't use the dimension information can just ignore. This would simplify the code a lot!

Reply:
I don't see a reason to support non-contiguous arrays. But wouldn't using `NDBuffer` exclusively mean that we lose the separation of the classes?

Reply:
> But wouldn't using `NDBuffer` exclusively mean that we lose the separation of the classes?

Yes, if we force `NDBuffer` to be contiguous always, we don’t need to differentiate between a n-dim object and a flat 1-dim object because components that don’t need the n-dim info can just ignore it (no need to convert or copy data). This means that all codecs become `ArrayArrayCodec` and we can remove all the Array-Bytes codec combinations.  

 

Now, if we really do need non-contiguous arrays, then the current design with all the Array-Bytes codec combinations is good because it makes conversion and data copying explicit. But if we are operating on contiguous arrays only, I don’t see the point of this added complexity.

Reply:
I am not sure I follow. I am advocating for separating bytes from ndarrays. That has been one of my design goals for the zarr-python 3 refactor. It brings a lot more clarity and invariants throughout the library. For example, a store will just not return a multi-dimensional array, a `bytes` codec will always decode into an ndarray, a `transpose` codec will always return an ndarray. Also, the separation into `ArrayArrayCodec`, `ArrayArrayCodec` and `ArrayBytesCodec` was deliberate. Turning all codecs into `ArrayArrayCodec` would be a step backwards, in my opinion.
I would even argue that separating bytes from ndarrays reduces complexity because each component can rely on the type interfaces instead of having to call `ensure_bytes` or `ensure_ndarray` everywhere.

Reply:
> I would even argue that separating bytes from ndarrays reduces complexity because each component can rely on the type interfaces instead of having to call ensure_bytes or ensure_ndarray everywhere.

If we force `NDBuffer` to be contiguous,  why would we need to call `ensure_bytes` or `ensure_ndarray`?

Reply:
> > I would even argue that separating bytes from ndarrays reduces complexity because each component can rely on the type interfaces instead of having to call ensure_bytes or ensure_ndarray everywhere.
> 
> If we force `NDBuffer` to be contiguous, why would we need to call `ensure_bytes` or `ensure_ndarray`?

To validate whether the input is a flat bytestream or a multi-dimenional array. If this information is not available through the type system anymore, there need to runtime checks to do this.

Reply:
I think we are talking past each other. By forcing `NDBuffer` to be contiguous, we are saying that `NDBuffer` is guaranteed to be a flat bytestream. Now, it might also have some additional information such as dtype and strides, but this can always be ignored. E.g., the `LocalStore` doesn’t care about dtype and strides, it just reads or writes the `NDBuffer` as one big flat bytestream.  

In fact, all our stores read and write contiguous blocks of memory, which makes it hard to think of a scenario that needs non-contiguous arrays. But please let me know if such a scenario exists.

Reply:
Should have a brief call to talk about this? Maybe that would be easier to resolve. I am free today afternoon (CEST).

My main point remains the separation and enforcement thereof. I understand that components can ignore dtype, stride etc. I would rather have it that they cannot even access this information or call methods such as `astype`, `transpose`, `reshape`. All of this ideally without runtime checks and through the type system.